File: scripts/packages.js
const childProcess = require("child_process");
const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const readline = require("readline");

const osPath = path.resolve(__dirname, "../.os");
const configPath = path.resolve(__dirname, "../lb/config");
const downloadedDebs = path.resolve(osPath, ".debs");
const DEB_REGEX = /[^:]+:\/\/.+\.deb$/g;

function download(url, file) {
  return new Promise((resolve, reject) => {
    childProcess.exec(
      `mkdir -p ${downloadedDebs} && curl -Lo ${path.resolve(
        downloadedDebs,
        file
      )} ${url}`,
      (err, stdout, stderr) => {
        if (err) return reject(err);
        if (stderr) {
          if (stderr.toString().indexOf("% Total") <= -1) {
            return reject(new Error(stderr.toString()));
          }
        }
        return resolve(stdout);
      }
    );
  });
}

function stdin() {
  const rl = readline.createInterface({
    input: process.stdin,
  });
  return new Promise((resolve, reject) => {
    try {
      let stdin = "";
      rl.on("line", (line) => (stdin += line + "\n"));
      rl.on("close", () => resolve(stdin));
    } catch (err) {
      return reject(err);
    }
  });
}

function removeFiles(filePaths) {
  filePaths.forEach((filePath) => {
    if (fs.existsSync(filePath)) {
      fs.rmSync(filePath);
    }
  });
}

function cleanPackages() {
  removeFiles([
    path.resolve(configPath, "package-lists/packages.list.chroot_install"),
    path.resolve(configPath, "package-lists/packages.list.chroot_live"),
    path.resolve(configPath, "package-lists/packages.list.binary"),
  ]);
}

function loadPackage(package) {
  if (DEB_REGEX.test(package.package)) {
    return download(
      package.package,
      `${crypto.createHash("md5").update(package.package).digest("hex")}.deb`
    ).then((result) => {
      console.log(`downloaded package ${package.package}`);
      return result;
    });
  } else {
    if (package.live && package.installed) {
      // list.chroot or list.chroot_install
      fs.appendFileSync(
        path.resolve(configPath, "package-lists/packages.list.chroot_install"),
        `${package.package.trim()}\n`
      );
    } else if (package.live && !package.installed) {
      // list.chroot_live
      fs.appendFileSync(
        path.resolve(configPath, "package-lists/packages.list.chroot_live"),
        `${package.package.trim()}\n`
      );
    } else if (!package.live && package.installed) {
      // add to calamares/modules/packages.conf and includes.installer/preseed.cfg
      fs.appendFileSync(
        path.resolve(configPath, "package-lists/packages.list.install"),
        `${package.package.trim()}\n`
      );
    }
    if (package.binary) {
      // list.binary
      fs.appendFileSync(
        path.resolve(configPath, "package-lists/packages.list.binary"),
        `${package.package.trim()}\n`
      );
    }
    console.log(`loaded package ${package.package}`);
  }
}

function loadPackages() {
  if (!fs.existsSync(configPath)) fs.mkdirSync(configPath);
  if (!fs.existsSync(path.resolve(configPath, "package-lists"))) {
    fs.mkdirSync(path.resolve(configPath, "package-lists"));
  }
  return stdin()
    .then((data) => {
      JSON.parse(data).forEach((package) => {
        if (typeof package !== "string") return loadPackage(package);
        return loadPackage({
          binary: false,
          installed: true,
          live: true,
          package: package,
        });
      });
    })
    .catch(console.error);
}

module.exports = {
  clean: cleanPackages,
  load: loadPackages,
};

File: src/loaders/packages.py
from util import download, shell, merge_dirs, mkdirs
from yaml import SafeLoader
import glob
import os
import re
import shutil
import yaml


class PackagesLoader:
    name = "packages"

    URI_REGEX = r"[^:]+:\/\/.+\.deb$"

    def __init__(self, deb):
        self.deb = deb

    async def load(self):
        packages = await self.get_packages()
        if not "packages" in self.deb.data:
            self.deb.data["packages"] = []
        if not "debs" in self.deb.data:
            self.deb.data["debs"] = []
        for p in packages:
            await self.load_package(Package(p))
        if not os.path.exists(
            os.path.join(self.deb.paths["lb"], "config-overrides/packages.chroot")
        ):
            os.makedirs(
                os.path.join(self.deb.paths["lb"], "config-overrides/packages.chroot")
            )
        for d in glob.glob(
            os.path.join(self.deb.paths["os"], ".debs/**/*.deb"),
            recursive=True,
            include_hidden=True,
        ):
            shell(
                "dpkg-name -s "
                + os.path.join(self.deb.paths["lb"], "config-overrides/packages.chroot")
                + " -o "
                + d
            )
        if (
            len(
                glob.glob(
                    os.path.join(
                        self.deb.paths["lb"], "config-overrides/packages.chroot"
                    ),
                    include_hidden=True,
                )
            )
            > 0
        ):
            await merge_dirs(
                os.path.join(self.deb.paths["lb"], "config-overrides/packages.chroot"),
                os.path.join(self.deb.paths["installer_install"], "debs"),
            )
            await merge_dirs(
                os.path.join(self.deb.paths["lb"], "config-overrides/packages.chroot"),
                os.path.join(self.deb.paths["chroot_install"], "debs"),
            )

    async def get_packages(self):
        packages = []
        package_names = set()
        for path in glob.glob(
            os.path.join(self.deb.paths["os"], "packages/**/*.yaml"),
            recursive=True,
            include_hidden=True,
        ):
            with open(path) as f:
                data = yaml.load(f, Loader=SafeLoader)
                if not type(data) is list:
                    continue
                for package in data:
                    package_name = None
                    if type(package) == str:
                        package_name = package
                    elif "package" in package:
                        package_name = package["package"]
                    if type(package_name) is str and package_name not in package_names:
                        package_names.add(package_name)
                        packages.append(package)
        return packages

    async def load_package(self, package):
        if not not re.match(self.URI_REGEX, package.package):
            self.deb.data["debs"].append(package)
            await mkdirs(os.path.join(self.deb.paths["os"], ".debs"))
            await download(
                package.package,
                os.path.join(
                    self.deb.paths["os"], ".debs", str(hash(package.package)) + ".deb"
                ),
            )
        elif package.package[-4:] == ".deb":
            if package.live or package.installed:
                await mkdirs(os.path.join(self.deb.paths["os"], ".debs"))
                shutil.copyfile(
                    os.path.join(self.deb.paths["os"], "packages", package.package),
                    os.path.join(
                        self.deb.paths["os"],
                        ".debs",
                        str(hash(package.package)) + ".deb",
                    ),
                )
        else:
            self.deb.data["packages"].append(package)
            if not os.path.exists(
                os.path.join(
                    self.deb.paths["lb"],
                    "config-overrides/package-lists",
                )
            ):
                os.makedirs(
                    os.path.join(
                        self.deb.paths["lb"],
                        "config-overrides/package-lists",
                    )
                )
            if package.live and package.installed:
                with open(
                    os.path.join(
                        self.deb.paths["lb"],
                        "config-overrides/package-lists/packages.list.chroot",
                    ),
                    "a",
                ) as f:
                    f.write(package.package + "\n")
            elif package.live and not package.installed:
                with open(
                    os.path.join(
                        self.deb.paths["lb"],
                        "config-overrides/package-lists/packages.list.chroot_live",
                    ),
                    "a",
                ) as f:
                    f.write(package.package + "\n")
            elif not package.live and package.installed:
                # add to calamares/modules/packages.conf and includes.installer/preseed.cfg
                # handled by overlays
                pass
            if package.binary:
                with open(
                    os.path.join(
                        self.deb.paths["lb"],
                        "config-overrides/package-lists/packages.list.binary",
                    ),
                    "a",
                ) as f:
                    f.write(package.package + "\n")


class Package:
    def __init__(self, package):
        if type(package) is not dict:
            package = {"package": package}
        self.live = package["live"] if "live" in package else True
        self.installed = package["installed"] if "installed" in package else True
        self.binary = package["binary"] if "binary" in package else False
        self.package = package["package"].strip()
